src/Language/Fixpoint/Solver.hs

```diff
@@ -348,27 +370,37 @@ simplifyResult res =
 --
 -- We require that relevant variables occur more than once, or that
 -- they occur in some other place than as an argument to @==@.
-simplifyKVar :: Expr -> Expr
-simplifyKVar = go
+simplifyKVarOccurrences :: Expr -> Expr
+simplifyKVarOccurrences = go
   where
     go (POr es) = POr $ map go es
     go (PExist bs e@(PAnd es)) =
-      let -- existential bindings that occur only once in the body of the
+      let -- Count occurrences of each variable
+          allOccurrences = L.group $ L.sort $ collectFreeVarOccurrences e
+
+          -- existential bindings that occur only once in the body of the
           -- existential
           singleOccurrenceBindings =
             filter isExistentialBinding $
-              concat $ filter occursExactlyOnce $
-                L.group $ L.sort $ collectFreeVarOccurrences e
+              concat $ filter occursExactlyOnce allOccurrences
+
+          -- existential bindings that occur more than once
+          multipleOccurrenceBindings =
+            filter isExistentialBinding $
+              map head $ filter occursMoreThanOnce allOccurrences
+
           isExistentialBinding = (`elem` map fst bs)
           occursExactlyOnce [_] = True
           occursExactlyOnce _   = False
+          occursMoreThanOnce (_:_:_) = True
+          occursMoreThanOnce _ = False
 
           esv = map (isUniqueEq singleOccurrenceBindings) es
           removed = mapMaybe fst esv
-          needed = singleOccurrenceBindings L.\\ removed
+          needed = (singleOccurrenceBindings L.\\ removed) ++ multipleOccurrenceBindings
           bs' = filter ((`elem` needed) . fst) bs
```

Comment:
👍 Oh, I see how this was wrong. Would this simpler version work?
```suggestion
          bs' = filter ((`notElem` removed) . fst) bs
```

Reply:
hm, that doesn't eliminate bindings that aren't used. Maybe your version is to be preferred.

Reply:
Yes, that doesn't remove the unused bindings AFAICT.